### ISSUE TITLE
Ordering of lists in the configuration file may not be respected

### ISSUE THREAD

**Issue by lelit** [docker/compose#2342]
I switched one of my projects to use version 1.5.0 with Python 3.4, and hit a subtle problem related to the _hash randomization_ feature introduced in Python 3.3.

One of the effects is that `docker-compose up` keeps recreating some of the services, even if nothing has been changed in their configuration.

The problem disappear executing `export PYTHONHASHSEED=0` (actually, any value except the default 'random') before running `docker-compose`.

I created a tiny example that demonstrate the problem: https://github.com/lelit/docker-compose-config-ordering-issue-demo

**Comment by dnephin** [comment]
There are some keys where we convert the list into a dict so that we can find overlaps between previous containers, but we never serialize it for comparison.  We use a json representation with sorted keys to make sure it's consistent.

From your example, it sounds like this is not a bug in compose itself, but an assumption in your tool, is that right?  If you use json with sorted keys it should work?

**Comment by lelit** [comment]
Right, as I said [here](https://github.com/lelit/docker-compose-config-ordering-issue-demo#conclusion) it's probably not a bug, but maybe a problem-waiting-to-appear: "my tool"/"doit" effectively serializes (or rather, computes an hash from) the options dictionary, to be able to answer the question "is the config changed since last run?".

As suggested on the referenced doit issue, it could use something like `pprint.pformat()` to obtain the equivalent of your json-with-sorted-keys suggestion, but that's not enough with compose: as you say, *even* lists get reordered, not just dictionary keys.

**Comment by dnephin** [comment]
https://github.com/docker/compose/blob/67dc90ec/compose/service.py#L496-L506 might be helpful for you. We already compute a hash based on the config

**Comment by lelit** [comment]
Does it take into account the fact that when extending the service, the ordering of lists changes?

**Comment by lelit** [comment]
As I said in the demo repo README, even docker-compose by itself is affected: running `docker-compose up` in series, without changing anything, it recreates some of the services...

**Comment by dnephin** [comment]
I can't reproduce it with docker-compose. I used your example config, and re-ran `docker-compose up` a bunch of times. I also tried changing the value of `PYTHONHASHSEED` to force it to break.  Every time it re-used the same container ("Starting" instead of "recreating"), which means the hash worked correctly.

Can you produce a sample that fails with docker-compose cli?

**Comment by lelit** [comment]
I will try.

**Comment by lelit** [comment]
Here it is:
```
$ git clone https://github.com/lelit/docker-compose-config-ordering-issue-demo.git
Cloning into 'docker-compose-config-ordering-issue-demo'...
remote: Counting objects: 24, done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 24 (delta 6), reused 22 (delta 4), pack-reused 0
Ricezione degli oggetti: 100% (24/24), 4.37 KiB | 0 bytes/s, done.
Risoluzione dei delta: 100% (6/6), done.
Checking connectivity... fatto.
$ cd docker-compose-config-ordering-issue-demo/
$ python3 -m venv env
$ source env/bin/activate
(env) $ pip install docker-compose
Downloading/unpacking docker-compose
  …
Downloading/unpacking docopt<0.7,>=0.6.1 (from docker-compose)
  …
Downloading/unpacking PyYAML<4,>=3.10 (from docker-compose)
  …
Downloading/unpacking requests<2.8,>=2.6.1 (from docker-compose)
  …
Downloading/unpacking texttable<0.9,>=0.8.1 (from docker-compose)
  …
Downloading/unpacking websocket-client<1.0,>=0.32.0 (from docker-compose)
  …
Downloading/unpacking docker-py<2,>=1.5.0 (from docker-compose)
  …
Downloading/unpacking dockerpty<0.4,>=0.3.4 (from docker-compose)
  …
Downloading/unpacking six<2,>=1.3.0 (from docker-compose)
  …
Downloading/unpacking jsonschema<3,>=2.5.1 (from docker-compose)
  …
Installing collected packages: docker-compose, docopt, PyYAML, requests, texttable, websocket-client, docker-py, dockerpty, six, jsonschema
  …
Successfully installed docker-compose docopt PyYAML requests texttable websocket-client docker-py dockerpty six jsonschema
Cleaning up...
(env) $ docker-compose up
Creating dockercomposeconfigorderingissuedemo_redis_1
Building web
Step 1 : FROM python:2.7
  …
Successfully built 79375d9b7da4
Creating dockercomposeconfigorderingissuedemo_web_1
Attaching to dockercomposeconfigorderingissuedemo_redis_1, dockercomposeconfigorderingissuedemo_web_1
  …
redis_1 | 1:M 09 Nov 18:27:05.092 * The server is now ready to accept connections on port 6379
web_1   |  * Restarting with stat
web_1   |  * Debugger is active!
web_1   |  * Debugger pin code: 205-067-129
^CGracefully stopping... (press Ctrl+C again to force)
Stopping dockercomposeconfigorderingissuedemo_web_1 ... done
Stopping dockercomposeconfigorderingissuedemo_redis_1 ... done
ERROR: 
Aborting.
(env) $ docker-compose up
Starting dockercomposeconfigorderingissuedemo_redis_1
Recreating dockercomposeconfigorderingissuedemo_web_1
Attaching to dockercomposeconfigorderingissuedemo_redis_1, dockercomposeconfigorderingissuedemo_web_1
  …
^CGracefully stopping... (press Ctrl+C again to force)
Stopping dockercomposeconfigorderingissuedemo_web_1 ... done
Stopping dockercomposeconfigorderingissuedemo_redis_1 ... done
ERROR: 
Aborting.
(env) $ docker-compose up
Starting dockercomposeconfigorderingissuedemo_redis_1
Recreating dockercomposeconfigorderingissuedemo_web_1
  …
^CGracefully stopping... (press Ctrl+C again to force)
Stopping dockercomposeconfigorderingissuedemo_web_1 ... done
Stopping dockercomposeconfigorderingissuedemo_redis_1 ... done
ERROR:
Aborting.
(env) $ export PYTHONHASHSEED=0
(env) $ docker-compose up
Starting dockercomposeconfigorderingissuedemo_redis_1
Recreating dockercomposeconfigorderingissuedemo_web_1
  …
^CGracefully stopping... (press Ctrl+C again to force)
Stopping dockercomposeconfigorderingissuedemo_web_1 ... done
Stopping dockercomposeconfigorderingissuedemo_redis_1 ... done
ERROR: 
Aborting.
(env) $ docker-compose up
Starting dockercomposeconfigorderingissuedemo_redis_1
Starting dockercomposeconfigorderingissuedemo_web_1
  …
^CGracefully stopping... (press Ctrl+C again to force)
Stopping dockercomposeconfigorderingissuedemo_web_1 ... done
Stopping dockercomposeconfigorderingissuedemo_redis_1 ... done
ERROR: 
Aborting.
$ unset PYTHONHASHSEED
(env) $ docker-compose up
Starting dockercomposeconfigorderingissuedemo_redis_1
Recreating dockercomposeconfigorderingissuedemo_web_1
  …
```

As you can see, when the randomization is in effect, the web service gets recreated each time. Disabling it with `PYTHONHASHSEED=0` it's simply started.

Does this happens to you?

**Comment by dnephin** [comment]
I was trying with the 1.5.0 release binary, which uses python2.7, so maybe it's a python3 issue.  Thanks for the example!

**Comment by lelit** [comment]
I snipped too much:

```
$ docker-compose --version
docker-compose version: 1.5.0
```

**Comment by lelit** [comment]
Yes, sure: the randomization is in effect by default under Py3. Under Py2, you must set `PYTHONSEEDHASH=random` to activate it.

**Comment by shin-** [comment]
Looks like this was fixed by #3334. Let me know if you're still seeing issues with recent versions.
